src/library/psi_library.jl

```diff
@@ -1149,3 +1149,415 @@ function build_two_zone_5_bus(; kwargs...)
+            name = "AC_interconnection",
+            available = true,
+            active_power_flow = 0.0,
+            reactive_power_flow = 0.0,
+            arc = get_component(Arc, main_sys, "Alder -> Alder_twin"),
+            r = 0.042,
+            x = 0.161,
+            b = (from = 0.022, to = 0.022),
+            rate = 1.75,
+            # For now, not binding
+            flow_limits = (from_to = 2.0, to_from = 2.0),
+            angle_limits = (min = -1.57079, max = 1.57079),
+            services = Vector{Service}[],
+            ext = Dict{String, Any}(),
+        )
+        PSY.add_component!(main_sys, new_ACLine)
+    end
+
+    for bat in get_components(GenericBattery, main_sys)
+        set_base_power!(bat, get_base_power(bat) * 10)
+    end
+
+    for r in get_components(
+        x -> get_prime_mover_type(x) == PrimeMovers.CP,
+        RenewableDispatch,
+        main_sys,
+    )
+        clear_services!(r)
+        remove_component!(main_sys, r)
+    end
+
+    for dev in get_components(RenewableFix, main_sys)
+        clear_services!(dev)
+    end
+
+    for dev in
+        get_components(x -> get_fuel(x) == ThermalFuels.NUCLEAR, ThermalStandard, main_sys)
+        clear_services!(dev)
+    end
+
+    for dev in get_components(HydroGen, main_sys)
+        clear_services!(dev)
+    end
+
+    bus_to_change = PSY.get_component(ACBus, main_sys, "Arne_twin")
+    PSY.set_bustype!(bus_to_change, PSY.ACBusTypes.PV)
+
+    # cost perturbation must be the same for each sub-system
+    rand_ix = 1
+    for g in get_components(
+        x -> x.prime_mover_type in [PrimeMovers.CT, PrimeMovers.CC],
+        ThermalStandard,
+        main_sys,
+    )
+        old_pwl_array = get_variable(get_operation_cost(g)) |> get_cost
+        new_pwl_array = similar(old_pwl_array)
+        for (ix, tup) in enumerate(old_pwl_array)
+            if ix ∈ [1, length(old_pwl_array)]
+                noise_val, rand_ix = iterate(COST_PERTURBATION_NOISE, rand_ix)
+                cost_noise = 50.0 * noise_val
+                new_pwl_array[ix] = ((tup[1] + cost_noise), tup[2])
+            else
+                try_again = true
+                while try_again
+                    noise_val, rand_ix = iterate(COST_PERTURBATION_NOISE, rand_ix)
+                    cost_noise = 50.0 * noise_val
+                    noise_val, rand_ix = iterate(COST_PERTURBATION_NOISE, rand_ix)
+                    power_noise = 0.01 * noise_val
+                    slope_previous =
+                        ((tup[1] + cost_noise) - old_pwl_array[ix - 1][1]) /
+                        ((tup[2] - power_noise) - old_pwl_array[ix - 1][2])
+                    slope_next =
+                        (-(tup[1] + cost_noise) + old_pwl_array[ix + 1][1]) /
+                        (-(tup[2] - power_noise) + old_pwl_array[ix + 1][2])
+                    new_pwl_array[ix] = ((tup[1] + cost_noise), (tup[2] - power_noise))
+                    try_again = slope_previous > slope_next
```

Comment:
Could we do some algebra to generate `cost_noise` and `power_noise` such that this always succeeds on the first try without harming the randomness?

Reply:
@josephmckinsey finds this problem interesting and seems to want to take it on

Reply:
to be resolved later, for now there is an added test.